src/Libraries/Microsoft.Extensions.Options.Contextual/Internal/ContextualOptionsFactory.cs

```diff
@@ -20,28 +21,29 @@ namespace Microsoft.Extensions.Options.Contextual;
 internal sealed class ContextualOptionsFactory<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TOptions> : IContextualOptionsFactory<TOptions>
     where TOptions : class
 {
-    private readonly IOptionsFactory<TOptions> _baseFactory;
     private readonly ILoadContextualOptions<TOptions>[] _loaders;
-    private readonly IPostConfigureContextualOptions<TOptions>[] _postConfigures;
-    private readonly IValidateContextualOptions<TOptions>[] _validations;
+
+    private readonly IConfigureOptions<TOptions>[] _setups;
+    private readonly IPostConfigureOptions<TOptions>[] _postConfigures;
+    private readonly IValidateOptions<TOptions>[] _validations;
 
     /// <summary>
     /// Initializes a new instance of the <see cref="ContextualOptionsFactory{TOptions}"/> class.
     /// </summary>
-    /// <param name="baseFactory">The factory to create instances of <typeparamref name="TOptions"/> with.</param>
     /// <param name="loaders">The configuration loaders to run.</param>
+    /// <param name="setups">The configuration actions to run.</param>
     /// <param name="postConfigures">The initialization actions to run.</param>
     /// <param name="validations">The validations to run.</param>
     public ContextualOptionsFactory(
-        IOptionsFactory<TOptions> baseFactory,
         IEnumerable<ILoadContextualOptions<TOptions>> loaders,
-        IEnumerable<IPostConfigureContextualOptions<TOptions>> postConfigures,
-        IEnumerable<IValidateContextualOptions<TOptions>> validations)
+        IEnumerable<IConfigureOptions<TOptions>> setups,
+        IEnumerable<IPostConfigureOptions<TOptions>> postConfigures,
+        IEnumerable<IValidateOptions<TOptions>> validations)
     {
-        _baseFactory = baseFactory;
         _loaders = loaders.ToArray();
-        _postConfigures = postConfigures.ToArray();
-        _validations = validations.ToArray();
+        _setups = setups as IConfigureOptions<TOptions>[] ?? new List<IConfigureOptions<TOptions>>(setups).ToArray();
```

Comment:
`new List<IConfigureOptions<TOptions>>(setups).ToArray()` is quite inefficient way allocate memory because we're reallocating it multiple times.

Wouldn't this work?
```suggestion
        _setups = setups as IConfigureOptions<TOptions>[] ?? setups.ToArray();
```

Reply:
resolved